gadopt/momentum_equation.py

```diff
@@ -18,217 +18,175 @@
+            if compressible_stress:
+                trial_tensor_jump -= (
+                    2 / 3 * identity * (dot(eq.n, trial) - dot(eq.n, bc["u"]))
+                )
+            trial_tensor_jump += transpose(trial_tensor_jump)
+            # Terms below are similar to the above terms for the DG dS integrals.
+            F += (
+                2
+                * sigma
+                * inner(outer(eq.n, eq.test), mu * trial_tensor_jump)
+                * eq.ds(bc_id)
+            )
+            F -= inner(mu * nabla_grad(eq.test), trial_tensor_jump) * eq.ds(bc_id)
+            F -= inner(outer(eq.n, eq.test), stress) * eq.ds(bc_id)
+
+        if "un" in bc:
+            trial_tensor_jump = outer(eq.n, eq.n) * (dot(eq.n, trial) - bc["un"])
+            if compressible_stress:
+                trial_tensor_jump -= 2 / 3 * identity * (dot(eq.n, trial) - bc["un"])
+            trial_tensor_jump += transpose(trial_tensor_jump)
+            # Terms below are similar to the above terms for the DG dS integrals.
+            F += (
+                2
+                * sigma
+                * inner(outer(eq.n, eq.test), mu * trial_tensor_jump)
+                * eq.ds(bc_id)
+            )
+            F -= inner(mu * nabla_grad(eq.test), trial_tensor_jump) * eq.ds(bc_id)
+            # We only keep the normal part of stress; the tangential part is assumed to
+            # be zero stress (i.e. free slip) or prescribed via "stress".
+            F -= dot(eq.n, eq.test) * dot(eq.n, dot(stress, eq.n)) * eq.ds(bc_id)
+
+        if "stress" in bc:  # a momentum flux, a.k.a. "force"
+            # Here we need only the third term because we assume jump_u = 0
+            # (u_ext = trial) and stress = n . (mu . stress_tensor).
+            F -= dot(eq.test, bc["stress"]) * eq.ds(bc_id)
+
+        if "normal_stress" in bc:
+            F += dot(eq.test, bc["normal_stress"] * eq.n) * eq.ds(bc_id)
+
+        # if "drag" in bc:  # (bottom) drag of the form tau = -drag trial |trial|
+        #     drag = bc["drag"]
+        #     unorm = pow(dot(trial_lagged, trial_lagged) + 1e-6, 0.5)
+        #     F += dot(eq.test, drag * unorm * trial) * eq.ds(bc_id)
+
+    return -F
+
+
+def pressure_gradient_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    assert normal_is_continuous(eq.test)
+
+    F = -dot(div(eq.test), eq.p) * eq.dx
+
+    # Integration by parts gives a natural condition on pressure (as part of a normal
+    # stress condition). For boundaries where the normal component of u is specified, we
+    # remove that condition.
+    for bc_id, bc in eq.bcs.items():
+        if "u" in bc or "un" in bc:
+            F += dot(eq.test, eq.n) * eq.p * eq.ds(bc_id)
+
+    return -F
+
+
+def divergence_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    assert normal_is_continuous(eq.u)
+
+    rho = eq.approximation.rho
```

Comment:
This seems to have changed from `rho_continuity`

Reply:
True, thank you. I realised only recently with Will that this was not appropriate. I corrected it in `update_approx` but forgot to push here.